### DISCUSSION TITLE
Feature Request: MonadTall layout with configurable Nmaster

### DISCUSSION THREAD

**Discussion by addy419** [qtile/qtile#2916] [Ideas]
Currently, Tile layout has configurable master pane but does not have focus/shift by direction and MonadTall has everything I need except increasing windows in master pane. Can this be configured?

**Comment by elParaguayo** [comment]
What do you mean by "increasing windows in master pane"? You can grow the window size in MonadTall with `lazy.layout.grow()`.

If that's not what you're after, can you explain a bit more clearly for me.

**Comment by addy419** [comment]
@elParaguayo by Nmaster, I mean the master length (no. of windows in the left column) Similar to the Tile layout http://docs.qtile.org/en/latest/manual/ref/layouts.html#tile but for MonadTall.

**Comment by helmecke** [comment]
@addy419 I'm in the progress of creating a personal MonadTall and MonadWide layout, which allows to increase master length.
It's still a little rough and removes features(indiviual window sizes) the original qtile xmonad layouts had.
I've also added a maximize layout modifier and the ability to change split orientation of windows in master pane.

You can find it [here](https://github.com/helmecke/qtile-config/blob/master/qtile/.config/qtile/xmonad.py).

**Reply by hamensman** [reply]
Looking nice. Would it be possible to set the master length on the fly? Meaning, I can press a key to make it one master, and press another for two masters?

Also, I think it would make sense if when there are just 2 windows, the windows in (your modified) MonadTall are vertical (they are horizontal) and the windows in MonadWide are horizontal (they are vertical),

**Reply by helmecke** [reply]
@hamensman You can use `lazy.layout.increase_nmaster()` and `lazy.layout.decrease_nmaster()` take a look at [my config](https://github.com/helmecke/qtile-config/blob/master/qtile/.config/qtile/config.py#L101-L102).

> Also, I think it would make sense if when there are just 2 windows, the windows in (your modified) MonadTall are vertical (they are horizontal) and the windows in MonadWide are horizontal (they are vertical),

With the default values the layouts should behave like xmonads `Tall` and `Wide` layouts. Here is an example Screenshot of `MonadTall`.
![image](https://user-images.githubusercontent.com/37172067/147855756-04458a6f-ff0e-46b9-ae42-ded304f68447.png)

**Reply by hamensman** [reply]
Thank you, @helmecke

When setting the margins, they appear a bit inconsistent when there is more than one master window:

![Screenshot_2022-02-04_13-19-50](https://user-images.githubusercontent.com/76445071/152539958-3d23ccf3-f83c-48d3-b01d-260643571628.png)
![Screenshot_2022-02-04_13-20-53](https://user-images.githubusercontent.com/76445071/152539965-d2412630-4652-43aa-a9a7-a61ca832f554.png)

Here, I set the layout margin to 5 in my config.

**Comment by hamensman** [comment]
@addy419 @elParaguayo @helmecke I also think it would be cool if you could configure columns of the slave window stack, like in awesomewm.

Here is an example with awesome's `tile` layout (equivalent to qtile's `monadtall`), with 2 master windows and 2 slave columns:
![Screenshot_2022-04-07_13-40-28](https://user-images.githubusercontent.com/76445071/162201044-f3e20002-ed18-4281-a09a-c859d8928849.png)

And this is an example of what's used in awesomewm `rc.lua` config file to produce this (from here: https://github.com/BrodieRobertson/dotfiles/blob/master/config/awesome/rc.lua):
```lua
-- Master and column manipulation
    awful.key({ modkey    }, "m",     function () awful.tag.incnmaster( 1, nil, true) end,
              {description = "increase the number of master clients", group = "layout"}),
    awful.key({ modkey, "Shift"   }, "m",     function () awful.tag.incnmaster(-1, nil, true) end,
              {description = "decrease the number of master clients", group = "layout"}),
    awful.key({ modkey,  }, "n",     function () awful.tag.incncol( 1, nil, true)    end,
              {description = "increase the number of columns", group = "layout"}),
    awful.key({ modkey, "Shift" }, "n",     function () awful.tag.incncol(-1, nil, true)    end,
              {description = "decrease the number of columns", group = "layout"}),
```
